### PR TITLE
Trim furiganas

### PR DESCRIPTION
**Pull request by jiru**
Rework the whole thing to allow partial furiganas on tokens. This breaks the
XML format and everything. For instance, お母さん is now a single token with
a reading for 母 (I removed CDATA for readability):

    <parse>
      <token>...</token>
      <token>お<reading furigana="かあ">母</reading>さん</token>
    </parse>

Instead of the previous:

    <parse>
      <furigana>...</furigana>
      <furigana>
         <token>お母さん</token>
         <kana>おかあさん</kana>
      </furigana>
    </parse>

This allows to set furiganas more independently from the tokens.

### CODE REVIEW

**Comment by allan-simon** [comment]
thanks for the PR, I will review it in details tomorrow but can you 

1. update the readme with that examples (I should have done it before, but as you're changing the output it's the good occasion i think to make the readme clearer about what it input/output ) 

2. add documentation in the new methods you've created

**Comment by jiru** [comment]
I updated the README and added some documentation to the new functions I added.

File: src/Furigana.cpp [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+        hiraganas.append(character, std::strlen(character));
+    }
+
+    return hiraganas;
+}
+
+static inline void remove_spaces(std::string &str) {
+   str.erase(std::remove_if(str.begin(), str.end(), (int(*)(int))std::isspace), str.end());
```
Comment (allan-simon) [inline src/Furigana.cpp]: for long lines can you indent  long lines / or lines with lot of parameters like this  (the original intent is 4 spaces by indenta

```
   str.erase(
   std::remove_if(
             str.begin(),
                 str.end(), (int(*)(int))std::isspace)
```

File: src/Furigana.cpp [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    const char *reading = (const char*) reading_start;
+    const char *kanjis_prev;
+    char kanji_char[5] = { '\0' }; /* 4 UTF-8 bytes + '\0' = 5 */
+    char  kana_char[5] = { '\0' }; /* 4 UTF-8 bytes + '\0' = 5 */
+
+    if (strcmp(kanjis, reading) == 0) {
+        return 0;
+    }
+
+    while (1) {
```
Comment (allan-simon) [inline src/Furigana.cpp]: it's c++ so you can use `true` instead of 1 :)

File: src/Furigana.cpp [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+/**
+ * Find how much we should trim the given reading and kanji strings.
+ * Returns the length of head and tail to trim.
+ */
+static void find_trim_boundaries(
+        std::string kanjis_std, std::string reading_std,
+        size_t *start_len, size_t *end_len)
+{
```
Comment (allan-simon) [inline src/Furigana.cpp]:
```
static void find_trim_boundaries(
    std::string kanjis_std,
    std::string reading_std,
    size_t *start_len,
    size_t *end_len
) {
```

(same for other function that have > 3 parameters

**Comment by allan-simon** [comment]
I've added some comments which are more about personnal style to match other files of this project
otherwise  thanks a lot for the documentation :)

**Comment by jiru** [comment]
I fixed the indentation.

**Comment by allan-simon** [comment]
actually for the indentation it follows this rule 

```
opening(  //  regardless  if  /  function signature/  function call , the opening token being the last thing on the line 
     parameter1,  // always one indentation , no tentative of trying to align with previous line etc.
     parameter2,
     paramater3,
)
```

the goal being to have something that will create as few of diff as possible in case of renaming of method name , or adding removing a  parameter to a function

**Comment by jiru** [comment]
I believe that’s the indentation style I followed, for functions that has more than 3 parameters. Where am I doing it wrong?

File: src/Server.cpp [hunk 4]
```diff
@@ -38,19 +37,29 @@ inline static void kana_output_xml(const char *kana, struct evbuffer *buffer) {
...
+inline static void cdata_output_xml(const char *string, struct evbuffer *buffer) {
+    evbuffer_add_printf(buffer, "<![CDATA[%s]]>", string);
 }
 
-inline static void furigana_output_xml_footer(struct evbuffer *buffer) {
-    evbuffer_add_printf(buffer, "</furigana>\n");
+inline static void reading_output_xml(std::pair<std::string, std::string> text,
+                                      struct evbuffer *buffer) {
+    evbuffer_add_printf(buffer, "<reading furigana=\"%s\"><![CDATA[%s]]></reading>",
+                        text.second.c_str(), text.first.c_str());
```
Comment (allan-simon) [inline src/Server.cpp]: here for example ,   
   
   * the  `buffer` variable should be on a separate line 
  *  each parameter should be on a line
  * `text.second.c_str() ` is aligned  with the method name (it should just be indented by 1)

File: src/Server.cpp [hunk 5]
```diff
@@ -38,19 +37,29 @@ inline static void kana_output_xml(const char *kana, struct evbuffer *buffer) {
...
-inline static void furigana_output_xml_header(struct evbuffer *buffer) {
-    evbuffer_add_printf(buffer, "<furigana>");
+inline static void cdata_output_xml(const char *string, struct evbuffer *buffer) {
+    evbuffer_add_printf(buffer, "<![CDATA[%s]]>", string);
 }
 
-inline static void furigana_output_xml_footer(struct evbuffer *buffer) {
-    evbuffer_add_printf(buffer, "</furigana>\n");
+inline static void reading_output_xml(std::pair<std::string, std::string> text,
+                                      struct evbuffer *buffer) {
```
Comment (allan-simon) [inline src/Server.cpp]: see comment below

File: src/Furigana.cpp [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    std::string readingString
+) {
+    size_t start_len, end_len;
+
+    remove_spaces(readingString);
+    readingString = this->katakana_to_hiragana(readingString);
+    find_trim_boundaries(this->katakana_to_hiragana(kanjisString),
+                         readingString,
+                         &start_len,
+                         &end_len);
```
Comment (allan-simon) [inline src/Furigana.cpp]: see comment on indentation

File: src/Server.cpp (hunk 4 above)
Comment (allan-simon) [inline src/Server.cpp]:
```
evbuffer_add_printf(
    buffer,
    "<reading furigana=\"%s\"><![CDATA[%s]]></reading>",
    text.second.c_str(),
    text.first.c_str()
);
```

**Comment by jiru** [comment]
Sorry, I was only looking at the function’s headers. I fixed the other parts. It should be okay, now.

**Comment by allan-simon** [comment]
great, thanks.

**Pull request merged**